File: sync/timeframe_filter.py
"""
Timeframe Filtering Logic
Part 20 - Trading Alerts SaaS

Filters SQLite rows based on timestamp matching for each timeframe.

Timeframe Categorization Rules (from architecture design):
- M5:  Minutes divisible by 5 (00, 05, 10, 15, ...)
- M15: Minutes divisible by 15 (00, 15, 30, 45)
- M30: Minutes divisible by 30 (00, 30)
- H1:  On the hour (minute == 0)
- H2:  Even hours (minute == 0 AND hour % 2 == 0)
- H4:  Hours divisible by 4 (minute == 0 AND hour % 4 == 0)
- H8:  Hours divisible by 8 (minute == 0 AND hour % 8 == 0)
- H12: Hours divisible by 12 (minute == 0 AND hour % 12 == 0)
- D1:  Midnight only (hour == 0 AND minute == 0)
"""

from datetime import datetime
from typing import List, Tuple, Dict, Any

# Timeframe divisors in minutes
TIMEFRAME_DIVISORS: Dict[str, int] = {
    "M5": 5,       # Minutes divisible by 5
    "M15": 15,     # Minutes divisible by 15
    "M30": 30,     # Minutes divisible by 30
    "H1": 60,      # On the hour
    "H2": 120,     # Every 2 hours
    "H4": 240,     # Every 4 hours
    "H8": 480,     # Every 8 hours
    "H12": 720,    # Every 12 hours
    "D1": 1440,    # Once per day (midnight)
}


def filter_by_timeframe(rows: List[Tuple], timeframe: str) -> List[Tuple]:
    """
    Filter rows to only include timestamps matching the timeframe.

    For M5: timestamps at :00, :05, :10, :15, :20, :25, :30, :35, :40, :45, :50, :55
    For M15: timestamps at :00, :15, :30, :45
    For M30: timestamps at :00, :30
    For H1: timestamps at :00:00 of each hour
    For H2: timestamps at :00:00 of even hours (0, 2, 4, 6, ...)
    For H4: timestamps at :00:00 of hours divisible by 4 (0, 4, 8, 12, 16, 20)
    For H8: timestamps at :00:00 of hours divisible by 8 (0, 8, 16)
    For H12: timestamps at :00:00 of hours divisible by 12 (0, 12)
    For D1: timestamps at 00:00:00 each day

    Args:
        rows: List of database row tuples where first element is Unix timestamp
        timeframe: Timeframe string (M5, M15, M30, H1, H2, H4, H8, H12, D1)

    Returns:
        Filtered list of rows matching the timeframe criteria
    """
    if timeframe not in TIMEFRAME_DIVISORS:
        return []

    filtered = []
    for row in rows:
        timestamp = row[0]  # First column is Unix timestamp
        dt = datetime.utcfromtimestamp(timestamp)

        if _matches_timeframe(dt, timeframe):
            filtered.append(row)

    return filtered


def _matches_timeframe(dt: datetime, timeframe: str) -> bool:
    """
    Check if a datetime matches the timeframe criteria.

    Args:
        dt: Datetime object to check
        timeframe: Timeframe string

    Returns:
        True if the datetime matches the timeframe criteria
    """
    if timeframe == "D1":
        # Daily: must be exactly midnight (00:00:00)
        return dt.hour == 0 and dt.minute == 0

    if timeframe == "H12":
        # H12: minute must be 0 and hour must be 0 or 12
        return dt.minute == 0 and dt.hour % 12 == 0

    if timeframe == "H8":
        # H8: minute must be 0 and hour must be divisible by 8 (0, 8, 16)
        return dt.minute == 0 and dt.hour % 8 == 0

    if timeframe == "H4":
        # H4: minute must be 0 and hour must be divisible by 4 (0, 4, 8, 12, 16, 20)
        return dt.minute == 0 and dt.hour % 4 == 0

    if timeframe == "H2":
        # H2: minute must be 0 and hour must be even
        return dt.minute == 0 and dt.hour % 2 == 0

    if timeframe == "H1":
        # H1: minute must be 0 (on the hour)
        return dt.minute == 0

    if timeframe == "M30":
        # M30: minute must be 0 or 30
        return dt.minute % 30 == 0

    if timeframe == "M15":
        # M15: minute must be 0, 15, 30, or 45
        return dt.minute % 15 == 0

    if timeframe == "M5":
        # M5: minute must be divisible by 5
        return dt.minute % 5 == 0

    return False


def get_latest_timeframe_timestamp(timeframe: str) -> int:
    """
    Get the most recent valid timestamp for a timeframe.

    Rounds down the current UTC time to the nearest valid timestamp
    for the specified timeframe.

    Args:
        timeframe: Timeframe string (M5, M15, M30, H1, H2, H4, H8, H12, D1)

    Returns:
        Unix timestamp of the most recent valid timeframe boundary
    """
    now = datetime.utcnow()

    if timeframe == "D1":
        # Round down to midnight
        aligned = now.replace(hour=0, minute=0, second=0, microsecond=0)

    elif timeframe == "H12":
        # Round down to 0 or 12 hours
        aligned_hour = (now.hour // 12) * 12
        aligned = now.replace(hour=aligned_hour, minute=0, second=0, microsecond=0)

    elif timeframe == "H8":
        # Round down to 0, 8, or 16 hours
        aligned_hour = (now.hour // 8) * 8
        aligned = now.replace(hour=aligned_hour, minute=0, second=0, microsecond=0)

    elif timeframe == "H4":
        # Round down to nearest 4-hour boundary
        aligned_hour = (now.hour // 4) * 4
        aligned = now.replace(hour=aligned_hour, minute=0, second=0, microsecond=0)

    elif timeframe == "H2":
        # Round down to nearest even hour
        aligned_hour = (now.hour // 2) * 2
        aligned = now.replace(hour=aligned_hour, minute=0, second=0, microsecond=0)

    elif timeframe == "H1":
        # Round down to current hour
        aligned = now.replace(minute=0, second=0, microsecond=0)

    elif timeframe == "M30":
        # Round down to 0 or 30 minutes
        aligned_minute = (now.minute // 30) * 30
        aligned = now.replace(minute=aligned_minute, second=0, microsecond=0)

    elif timeframe == "M15":
        # Round down to 0, 15, 30, or 45 minutes
        aligned_minute = (now.minute // 15) * 15
        aligned = now.replace(minute=aligned_minute, second=0, microsecond=0)

    elif timeframe == "M5":
        # Round down to nearest 5-minute boundary
        aligned_minute = (now.minute // 5) * 5
        aligned = now.replace(minute=aligned_minute, second=0, microsecond=0)

    else:
        # Default to M5 if unknown timeframe
        aligned_minute = (now.minute // 5) * 5
        aligned = now.replace(minute=aligned_minute, second=0, microsecond=0)

    return int(aligned.timestamp())


def get_timeframe_duration_seconds(timeframe: str) -> int:
    """
    Get the duration of a timeframe in seconds.

    Args:
        timeframe: Timeframe string

    Returns:
        Duration in seconds
    """
    divisor = TIMEFRAME_DIVISORS.get(timeframe, 5)
    return divisor * 60


def categorize_timestamp(timestamp: int) -> List[str]:
    """
    Determine which timeframes a timestamp belongs to.

    A timestamp can belong to multiple timeframes. For example,
    midnight belongs to all timeframes (M5, M15, M30, H1, H2, H4, H8, H12, D1).

    Args:
        timestamp: Unix timestamp

    Returns:
        List of timeframe strings this timestamp belongs to
    """
    dt = datetime.utcfromtimestamp(timestamp)
    matching_timeframes = []

    for timeframe in TIMEFRAME_DIVISORS:
        if _matches_timeframe(dt, timeframe):
            matching_timeframes.append(timeframe)

    return matching_timeframes

File: lib/indicators/timeframe-filter.ts
/**
 * Part 20 - Timeframe Filter Utility
 *
 * Filters indicator data rows based on timeframe boundaries.
 * Used by the sync script and API to categorize data into appropriate timeframes.
 *
 * @see docs/sqlite-and-mt5service/part-20-architecture-design.md Section 4.3
 */

/**
 * Timeframe divisors in minutes
 *
 * - M5: timestamps at 5-minute boundaries
 * - M15: timestamps at 15-minute boundaries
 * - M30: timestamps at 30-minute boundaries
 * - H1: timestamps on the hour
 * - H2: timestamps on even hours
 * - H4: timestamps on hours divisible by 4
 * - H8: timestamps on hours divisible by 8
 * - H12: timestamps on hours divisible by 12
 * - D1: timestamps at midnight only
 */
export const TIMEFRAME_DIVISORS: Record<string, number> = {
  M5: 5, // Minutes divisible by 5
  M15: 15, // Minutes divisible by 15
  M30: 30, // Minutes divisible by 30
  H1: 60, // On the hour (60 minutes)
  H2: 120, // Even hours (120 minutes)
  H4: 240, // Hours divisible by 4 (240 minutes)
  H8: 480, // Hours divisible by 8 (480 minutes)
  H12: 720, // Hours divisible by 12 (720 minutes)
  D1: 1440, // Midnight only (1440 minutes = 1 day)
};

/**
 * Valid timeframes
 */
export const VALID_TIMEFRAMES = [
  'M5',
  'M15',
  'M30',
  'H1',
  'H2',
  'H4',
  'H8',
  'H12',
  'D1',
] as const;

export type Timeframe = (typeof VALID_TIMEFRAMES)[number];

/**
 * Generic row interface for filtering
 */
export interface FilterableRow {
  timestamp: number; // Unix timestamp in seconds
  [key: string]: unknown;
}

/**
 * Filter rows based on timeframe boundaries
 *
 * @param rows - Array of rows with timestamp field (Unix timestamp in seconds)
 * @param timeframe - Target timeframe (M5, M15, M30, H1, H2, H4, H8, H12, D1)
 * @returns Filtered rows matching the timeframe boundaries
 *
 * @example
 * const rows = [
 *   { timestamp: 1704067200, ... }, // 00:00
 *   { timestamp: 1704067500, ... }, // 00:05
 *   { timestamp: 1704067800, ... }, // 00:10
 * ];
 *
 * filterByTimeframe(rows, 'M5');  // Returns all rows (all on 5-min boundaries)
 * filterByTimeframe(rows, 'M15'); // Returns [{ timestamp: 1704067200 }] (only 00:00 and 00:15, etc.)
 */
export function filterByTimeframe<T extends FilterableRow>(
  rows: T[],
  timeframe: string
): T[] {
  const upperTimeframe = timeframe.toUpperCase();
  const divisor = TIMEFRAME_DIVISORS[upperTimeframe];

  if (divisor === undefined) {
    console.warn(`[TimeframeFilter] Unknown timeframe: ${timeframe}`);
    return [];
  }

  return rows.filter((row) => {
    const timestamp = row.timestamp;
    const date = new Date(timestamp * 1000); // Convert to milliseconds

    // Get hours and minutes in UTC
    const hours = date.getUTCHours();
    const minutes = date.getUTCMinutes();
    const seconds = date.getUTCSeconds();

    // Calculate minutes since midnight
    const minutesSinceMidnight = hours * 60 + minutes;

    // Handle D1 (Daily): must be exactly midnight
    if (upperTimeframe === 'D1') {
      return hours === 0 && minutes === 0 && seconds === 0;
    }

    // Handle hourly timeframes (H1, H2, H4, H8, H12)
    if (upperTimeframe.startsWith('H')) {
      // Minute and seconds must be 0
      if (minutes !== 0 || seconds !== 0) {
        return false;
      }
      // Check if hours align with divisor
      return minutesSinceMidnight % divisor === 0;
    }

    // Handle minute timeframes (M5, M15, M30)
    // Seconds should be 0 and minutes should be divisible
    if (seconds !== 0) {
      return false;
    }
    return minutes % divisor === 0;
  });
}

/**
 * Check if a timestamp matches a timeframe boundary
 *
 * @param timestamp - Unix timestamp in seconds
 * @param timeframe - Target timeframe
 * @returns true if timestamp matches the timeframe boundary
 */
export function isTimeframeBoundary(
  timestamp: number,
  timeframe: string
): boolean {
  const row = { timestamp } as FilterableRow;
  const result = filterByTimeframe([row], timeframe);
  return result.length > 0;
}

/**
 * Get the latest valid timestamp for a timeframe
 *
 * @param timeframe - Target timeframe
 * @param referenceTime - Reference time (defaults to now)
 * @returns Unix timestamp aligned to the timeframe
 */
export function getLatestTimeframeTimestamp(
  timeframe: string,
  referenceTime?: Date
): number {
  const now = referenceTime || new Date();
  const upperTimeframe = timeframe.toUpperCase();
  const divisor = TIMEFRAME_DIVISORS[upperTimeframe];

  if (divisor === undefined) {
    throw new Error(`Unknown timeframe: ${timeframe}`);
  }

  // Start with a copy of the reference time
  const aligned = new Date(now);

  // Zero out seconds and milliseconds
  aligned.setUTCSeconds(0, 0);

  if (upperTimeframe === 'D1') {
    // Daily: align to midnight
    aligned.setUTCHours(0, 0, 0, 0);
  } else if (upperTimeframe.startsWith('H')) {
    // Hourly: align to hour boundary
    aligned.setUTCMinutes(0, 0, 0);
    const hours = aligned.getUTCHours();
    const hoursMultiple = divisor / 60;
    const alignedHour = Math.floor(hours / hoursMultiple) * hoursMultiple;
    aligned.setUTCHours(alignedHour);
  } else {
    // Minute: align to minute boundary
    const minutes = aligned.getUTCMinutes();
    const alignedMinute = Math.floor(minutes / divisor) * divisor;
    aligned.setUTCMinutes(alignedMinute);
  }

  return Math.floor(aligned.getTime() / 1000);
}

/**
 * Get the next timeframe boundary timestamp
 *
 * @param timeframe - Target timeframe
 * @param referenceTime - Reference time (defaults to now)
 * @returns Unix timestamp of the next boundary
 */
export function getNextTimeframeBoundary(
  timeframe: string,
  referenceTime?: Date
): number {
  const current = getLatestTimeframeTimestamp(timeframe, referenceTime);
  const divisor = TIMEFRAME_DIVISORS[timeframe.toUpperCase()];

  // Add the divisor in seconds
  return current + divisor * 60;
}

/**
 * Get the number of bars expected between two timestamps for a timeframe
 *
 * @param startTimestamp - Start Unix timestamp
 * @param endTimestamp - End Unix timestamp
 * @param timeframe - Target timeframe
 * @returns Expected number of bars
 */
export function getExpectedBarCount(
  startTimestamp: number,
  endTimestamp: number,
  timeframe: string
): number {
  const divisor = TIMEFRAME_DIVISORS[timeframe.toUpperCase()];

  if (!divisor) {
    return 0;
  }

  const diffSeconds = endTimestamp - startTimestamp;
  const divisorSeconds = divisor * 60;

  return Math.floor(diffSeconds / divisorSeconds) + 1;
}
